src/analysis/forward.jl

```diff
@@ -1,5 +1,65 @@
 using Core.Compiler: StmtInfo, ArgInfo, CallMeta, AbsIntState
 
+if VERSION >= v"1.12.0-DEV.1268"
+
+using Core.Compiler: Future
+
+function fwd_abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
+        arginfo::ArgInfo, si::StmtInfo, sv::AbsIntState, primal_call::Future{CallMeta})
+    if f === ChainRulesCore.frule
+        # TODO: Currently, we don't have any termination analysis for the non-stratified
+        # forward analysis, so bail out here.
+        return Future{Union{CallMeta, Nothing}}(nothing)
+    end
+
+    nargs = length(arginfo.argtypes)-1
+    frule_preargtypes = Any[Const(ChainRulesCore.frule), Tuple{Nothing,Vararg{Any,nargs}}]
+    frule_argtypes = append!(frule_preargtypes, arginfo.argtypes)
+    frule_atype = CC.argtypes_to_type(frule_argtypes)
+
+    local frule_call::Future{CallMeta}
+    local result::Future{Union{CallMeta, Nothing}} = Future{Union{CallMeta, Nothing}}()
+    function make_progress(_, sv)
```

Comment:
@vtjnash, please confirm that this is the intended way to use this.

Reply:
You seem to have closure captured interp instead of using the argument? The interp struct is commonly quite large, so that can increase memory usage quite a bit

Reply:
The argument is the wrong interp

Reply:
Yes, this way of defining a `make_progress` seems fine. There isn't really one right answer about how to code this, so base itself already uses probably 3 or 4 different patterns, depending on what kept the original code control flow seemed least distorted. I hadn't used the `@isdefined` trick, but it is essentially equivalent to the nextstate pattern I'd used for manual stackless state machine conversion

Reply:
I don't think capturing the interp will give you the behavior you want. I think you might need to mutate `sv` instead of re-using `sv` with different `interp`

Reply:
I'm not re-using `sv` with a different `interp`. `sv` here is an `IRInterpretationState`, which doesn't have an `interp` argument, so when the callback later gets scheduled, there's just some random interp in there.

Reply:
sv claims to be a `AbsIntState` here? For `IRInterpretationState` currently it passes the interp here that was originally used to construct the `IRInterpretationState`, since everything is on the stack there and doesn't handle recursion

I think the behavior here is also probably fine, but that no other callback will be using the right interp, since none other are expecting the interp to be different from the one used to allocate the state object

Reply:
> and doesn't handle recursion

Doesn't handle recursion in `Base` ;).

Reply:
Fair, I know almost nothing about this code, so I am reviewing without really knowing how this integrates. The current implementation in Base would potentially break here: https://github.com/JuliaLang/julia/blob/be401635fe02b28ce994e2e3cae0733d101f8927/base/compiler/ssair/irinterp.jl#L154
since it was not tracking if the return type changed to reschedule this instruction if it became part of cycle (I believe it should detect and `@assert` though if that attempts to happen)